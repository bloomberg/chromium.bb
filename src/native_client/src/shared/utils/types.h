/*
 * Copyright 2009 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#ifndef NATIVE_CLIENT_SRC_SHARED_UTILS_TYPES_H__
#define NATIVE_CLIENT_SRC_SHARED_UTILS_TYPES_H__

#include "native_client/src/include/build_config.h"
#include "native_client/src/include/portability.h"

EXTERN_C_BEGIN

/* Note: the names Bool, FALSE and TRUE are defined in some directories,
 * which conflict with native_client/src/include/portability.h. Hence,
 * this file has not been encorporated into that header file.
 */

/* Define an implementation for boolean values. */
#if NACL_WINDOWS
typedef BOOL Bool;
#else
typedef enum {
  FALSE,
  TRUE
} Bool;
#endif

EXTERN_C_END

#endif   /* NATIVE_CLIENT_SRC_SHARED_UTILS_TYPES_H__ */
