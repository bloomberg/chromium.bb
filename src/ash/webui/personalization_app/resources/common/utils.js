// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview Utility functions to be shared between trusted and untrusted
 * code.
 */

/**
 * @param {string} layout
 * @return {ash.personalizationApp.mojom.WallpaperLayout}
 */
export function getWallpaperLayoutEnum(layout) {
  switch (layout) {
    case 'FILL':
      return ash.personalizationApp.mojom.WallpaperLayout.kCenterCropped;
    case 'CENTER': // fall through
    default:
      return ash.personalizationApp.mojom.WallpaperLayout.kCenter;
  }
}

/**
 * Checks if argument is an array with non-zero length.
 * @param {?Object} maybeArray
 * @return {boolean}
 */
export function isNonEmptyArray(maybeArray) {
  return Array.isArray(maybeArray) && maybeArray.length > 0;
}

/**
 * Checks if argument is null or is an array.
 * @param {?Array} maybeNullOrArray
 * @return {boolean}
 */
export function isNullOrArray(maybeNullOrArray) {
  return maybeNullOrArray === null || Array.isArray(maybeNullOrArray);
}

/**
 * Checks if argument is null or is a number.
 * @param {?number} maybeNullOrNumber
 * @return {boolean}
 */
export function isNullOrNumber(maybeNullOrNumber) {
  return maybeNullOrNumber === null || typeof maybeNullOrNumber === 'number';
}

/**
 * Attach a listener to a child element onload function. Returns a promise
 * that resolves when that child element is loaded.
 * @param {*} element A polymer element
 * @param {!string} id Id of the child element.
 * @param {function(*, function(...*): void, !Array=): void}
 *     afterNextRender callback for first render of element.
 * @return {!Promise<!HTMLElement>}
 */
export function promisifyOnload(element, id, afterNextRender) {
  const promise = new Promise((resolve) => {
    function readyCallback() {
      const child = element.shadowRoot.getElementById(id);
      child.onload = () => resolve(child);
    }
    afterNextRender(element, readyCallback);
  });
  return promise;
}

/**
 * Returns true if this event is a user action to select an item.
 * @param {!Event} event
 * @return {boolean}
 */
export function isSelectionEvent(event) {
  return (event instanceof MouseEvent && event.type === 'click') ||
      (event instanceof KeyboardEvent && event.key === 'Enter');
}

/**
 * Sets a css variable to control the animation delay.
 * @param {number} index
 * @return {string}
 */
export function getLoadingPlaceholderAnimationDelay(index) {
  return `--animation-delay: ${index * 83}ms;`;
}

/**
 * Returns the number of grid items to render per row given the current inner
 * width of the |window|.
 * @return {number}
 */
export function getNumberOfGridItemsPerRow() {
  return window.innerWidth > 688 ? 4 : 3;
}

/**
 * Normalizes the given |key| for RTL.
 * @param {string} key
 * @param {boolean} isRTL
 * @return {string}
 */
export function normalizeKeyForRTL(key, isRTL) {
  if (isRTL) {
    if (key === 'ArrowLeft') {
      return 'ArrowRight';
    }
    if (key === 'ArrowRight') {
      return 'ArrowLeft';
    }
  }
  return key;
}
