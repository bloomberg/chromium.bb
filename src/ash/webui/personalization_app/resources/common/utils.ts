// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview Utility functions to be shared between trusted and untrusted
 * code.
 */

/**
 * Checks if argument is an array with non-zero length.
 */
export function isNonEmptyArray(maybeArray: unknown): maybeArray is unknown[] {
  return Array.isArray(maybeArray) && maybeArray.length > 0;
}

/**
 * Checks if argument is null or is an array.
 */
export function isNullOrArray(maybeNullOrArray: unknown):
    maybeNullOrArray is unknown[]|null {
  return maybeNullOrArray === null || Array.isArray(maybeNullOrArray);
}

/** Checks if argument is null or is a bigint. */
export function isNullOrBigint(maybeNullOrBigint: unknown):
    maybeNullOrBigint is bigint|null {
  return maybeNullOrBigint === null || typeof maybeNullOrBigint === 'bigint';
}

/**
 * Checks if argument is null or is a number.
 */
export function isNullOrNumber(maybeNullOrNumber: unknown):
    maybeNullOrNumber is number|null {
  return maybeNullOrNumber === null || typeof maybeNullOrNumber === 'number';
}

/**
 * Attach a listener to a child element onload function. Returns a promise
 * that resolves when that child element is loaded.
 * `id` Id of the child element.
 * `afterNextRender` callback for first render of element.
 */
export function promisifyOnload(
    element: HTMLElement, id: string,
    afterNextRender: (element: HTMLElement, callback: (...args: any) => void) =>
        void): Promise<HTMLElement> {
  const promise = new Promise<HTMLElement>((resolve) => {
    function readyCallback() {
      if (!element.shadowRoot) {
        return;
      }
      const child = element.shadowRoot.getElementById(id);
      if (!child) {
        return;
      }
      child.onload = () => resolve(child);
    }
    afterNextRender(element, readyCallback);
  });
  return promise;
}


type WallpaperSelectionEvent =
    MouseEvent&{type: 'click'}|KeyboardEvent&{key: 'Enter'};
/**
 * Returns true if this event is a user action to select an item.
 */
export function isSelectionEvent(event: Event):
    event is WallpaperSelectionEvent {
  return (event instanceof MouseEvent && event.type === 'click') ||
      (event instanceof KeyboardEvent && event.key === 'Enter');
}

/**
 * Sets a css variable to control the animation delay.
 */
export function getLoadingPlaceholderAnimationDelay(index: number): string {
  return `--animation-delay: ${index * 83}ms;`;
}

/**
 * Returns the number of grid items to render per row given the current inner
 * width of the |window|.
 */
export function getNumberOfGridItemsPerRow(): number {
  return window.innerWidth > 688 ? 4 : 3;
}

/**
 * Normalizes the given |key| for RTL.
 */
export function normalizeKeyForRTL(key: string, isRTL: boolean): string {
  if (isRTL) {
    if (key === 'ArrowLeft') {
      return 'ArrowRight';
    }
    if (key === 'ArrowRight') {
      return 'ArrowLeft';
    }
  }
  return key;
}
