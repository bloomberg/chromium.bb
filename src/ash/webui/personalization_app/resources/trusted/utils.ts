// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * @fileoverview Utility functions to be used in trusted code.
 */

import {FilePath} from 'chrome://resources/mojo/mojo/public/mojom/base/file_path.mojom-webui.js';

import {WallpaperImage, WallpaperLayout} from '../trusted/personalization_app.mojom-webui.js';

export function isWallpaperImage(obj: any): obj is WallpaperImage {
  return typeof obj?.assetId === 'bigint';
}

export function isFilePath(obj: any): obj is FilePath {
  return typeof obj?.path === 'string' && obj.path;
}

/**
 * Convert a string layout value to the corresponding enum.
 */
export function getWallpaperLayoutEnum(layout: string): WallpaperLayout {
  switch (layout) {
    case 'FILL':
      return WallpaperLayout.kCenterCropped;
    case 'CENTER':  // fall through
    default:
      return WallpaperLayout.kCenter;
  }
}

/**
 * Checks if argument is a string with non-zero length.
 */
export function isNonEmptyString(maybeString: unknown): maybeString is string {
  return typeof maybeString === 'string' && maybeString.length > 0;
}

/**
 * Wallpaper images sometimes have a resolution suffix appended to the end of
 * the image. This is typically to fetch a high resolution image to show as the
 * user's wallpaper. We do not want the full resolution here, so remove the
 * suffix to get a 512x512 preview.
 * TODO(b/186807814) support different resolution parameters here.
 */
export function removeHighResolutionSuffix(url: string): string {
  return url.replace(/=w\d+$/, '');
}

/**
 * Returns whether the given URL starts with http:// or https://.
 */
export function hasHttpScheme(url: string): boolean {
  return url.startsWith('http://') || url.startsWith('https://');
}
