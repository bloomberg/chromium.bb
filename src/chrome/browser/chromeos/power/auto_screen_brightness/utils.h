// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef CHROME_BROWSER_CHROMEOS_POWER_AUTO_SCREEN_BRIGHTNESS_UTILS_H_
#define CHROME_BROWSER_CHROMEOS_POWER_AUTO_SCREEN_BRIGHTNESS_UTILS_H_

#include <string>
#include <vector>

#include "base/containers/ring_buffer.h"
#include "base/time/time.h"
#include "chrome/browser/chromeos/power/auto_screen_brightness/trainer.h"

namespace chromeos {
namespace power {
namespace auto_screen_brightness {

// Replaces |curve| with the parsed curve data from |data|, returning true if
// successful. Correct formatting in |data| should be 1 row per
// (<ambient_light>, <brightness>) mapping. Ambient light should be a double and
// brightness should be a double.
bool CurveFromString(const std::string& data, BrightnessCurve* const curve);

// Converts |curve| to a string. An empty string will be returned if |curve| is
// empty.
std::string CurveToString(const BrightnessCurve& curve);

struct AmbientLightSample {
  int lux;
  base::TimeTicks sample_time;
};

// Calculates average ambient light over the most recent |num_recent| samples.
// |num_recent| has to be no larger than the capacity of the buffer, but it can
// be greater than the number of samples stored. If |num_recent| is greater than
// the number of samples stored, it returns average of stored samples. If
// |num_recent| is negative, it returns average over all stored samples. If
// input |ambient_light_values| is empty, it returns 0.
// TODO(jiameng): this is the very basic version of averaging, we need to change
// it to allow for time decay.
template <size_t size>
double AverageAmbient(
    const base::RingBuffer<AmbientLightSample, size>& ambient_light_values,
    int num_recent) {
  DCHECK_LE(num_recent, static_cast<int>(ambient_light_values.BufferSize()));
  if (ambient_light_values.CurrentIndex() == 0)
    return 0;

  double sum = 0;
  int count = 0;
  const int used_num_recent =
      num_recent < 0 ? ambient_light_values.BufferSize() : num_recent;
  for (auto it = ambient_light_values.End(); it && count < used_num_recent;
       --it) {
    const auto& ambient_sample = **it;
    sum += ambient_sample.lux;
    ++count;
  }
  return sum / count;
}

}  // namespace auto_screen_brightness
}  // namespace power
}  // namespace chromeos

#endif  // CHROME_BROWSER_CHROMEOS_POWER_AUTO_SCREEN_BRIGHTNESS_UTILS_H_
