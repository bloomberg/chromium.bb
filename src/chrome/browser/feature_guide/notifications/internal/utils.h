// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#ifndef CHROME_BROWSER_FEATURE_GUIDE_NOTIFICATIONS_INTERNAL_UTILS_H_
#define CHROME_BROWSER_FEATURE_GUIDE_NOTIFICATIONS_INTERNAL_UTILS_H_

#include <map>
#include <string>

#include "base/feature_list.h"
#include "build/build_config.h"
#include "chrome/browser/feature_guide/notifications/feature_type.h"

namespace feature_guide {

// Serialize a given FeatureType to notification custom data.
void FeatureToCustomData(FeatureType feature,
                         std::map<std::string, std::string>* custom_data);

// Deserialize the FeatureType from notification custom data.
FeatureType FeatureFromCustomData(
    const std::map<std::string, std::string>& custom_data);

// Get a fixed notification ID for the given feature.
std::string NotificationIdForFeature(FeatureType feature);

// Returns the feature type from the notification ID.
FeatureType NotificationIdToFeature(const std::string& notification_id);

#if defined(OS_ANDROID)
// Returns the notification IPH feature for the given feature.
base::Feature GetNotificationIphFeatureForFeature(FeatureType& feature);
#endif

}  // namespace feature_guide

#endif  // CHROME_BROWSER_FEATURE_GUIDE_NOTIFICATIONS_INTERNAL_UTILS_H_
