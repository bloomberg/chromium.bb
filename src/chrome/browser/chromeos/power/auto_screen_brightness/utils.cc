// Copyright 2018 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/chromeos/power/auto_screen_brightness/utils.h"

#include "base/strings/string_number_conversions.h"
#include "base/strings/string_split.h"
#include "base/strings/string_util.h"

namespace chromeos {
namespace power {
namespace auto_screen_brightness {

bool CurveFromString(const std::string& data, BrightnessCurve* const curve) {
  DCHECK(curve);
  curve->clear();
  if (data.empty())
    return true;

  base::StringPairs key_value_pairs;
  if (!base::SplitStringIntoKeyValuePairs(data, ',', '\n', &key_value_pairs)) {
    LOG(ERROR) << "Ill-formatted curve";
    return false;
  }

  for (base::StringPairs::iterator it = key_value_pairs.begin();
       it != key_value_pairs.end(); ++it) {
    double lux;
    if (!base::StringToDouble(it->first, &lux)) {
      LOG(ERROR) << "Ill-formatted lux";
      curve->clear();
      return false;
    }

    double brightness;
    if (!base::StringToDouble(it->second, &brightness)) {
      LOG(ERROR) << "Ill-formatted brightness";
      curve->clear();
      return false;
    }

    curve->push_back(std::make_pair(lux, brightness));
  }
  return true;
}

std::string CurveToString(const BrightnessCurve& curve) {
  if (curve.empty())
    return "";

  std::vector<std::string> rows;
  for (const auto& kv : curve) {
    rows.push_back(base::JoinString(
        {base::NumberToString(kv.first), base::NumberToString(kv.second)},
        ","));
  }

  return base::JoinString(rows, "\n");
}

}  // namespace auto_screen_brightness
}  // namespace power
}  // namespace chromeos
