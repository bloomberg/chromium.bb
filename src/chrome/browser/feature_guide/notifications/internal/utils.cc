// Copyright 2021 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/feature_guide/notifications/internal/utils.h"

#include "base/notreached.h"
#include "base/strings/string_number_conversions.h"
#include "build/build_config.h"
#include "components/feature_engagement/public/feature_list.h"

namespace feature_guide {
namespace {

constexpr char kCustomDataKeyForFeatureType[] = "feature_type";

constexpr char kNotificationIdDefaultBroser[] = "feature_guide_default_browser";
constexpr char kNotificationIdSignIn[] = "feature_guide_sign_in";
constexpr char kNotificationIdIncognitoTab[] = "feature_guide_incognito_tab";
constexpr char kNotificationIdVoiceSearch[] = "feature_guide_voice_search";
constexpr char kNotificationIdNTPSuggestionCard[] =
    "feature_guide_ntp_suggestion_card";

}  // namespace

void FeatureToCustomData(FeatureType feature,
                         std::map<std::string, std::string>* custom_data) {
  custom_data->emplace(kCustomDataKeyForFeatureType,
                       base::NumberToString(static_cast<int>(feature)));
}

FeatureType FeatureFromCustomData(
    const std::map<std::string, std::string>& custom_data) {
  int parsed_value = 0;
  std::string feature_string = custom_data.at(kCustomDataKeyForFeatureType);
  if (!base::StringToInt(feature_string, &parsed_value))
    return FeatureType::kInvalid;

  return static_cast<FeatureType>(parsed_value);
}

std::string NotificationIdForFeature(FeatureType feature) {
  switch (feature) {
    case FeatureType::kDefaultBrowser:
      return kNotificationIdDefaultBroser;
    case FeatureType::kSignIn:
      return kNotificationIdSignIn;
    case FeatureType::kIncognitoTab:
      return kNotificationIdIncognitoTab;
    case FeatureType::kNTPSuggestionCard:
      return kNotificationIdNTPSuggestionCard;
    case FeatureType::kVoiceSearch:
      return kNotificationIdVoiceSearch;
    default:
      NOTREACHED();
      return std::string();
  }
}

FeatureType NotificationIdToFeature(const std::string& notification_id) {
  if (notification_id == kNotificationIdDefaultBroser) {
    return FeatureType::kDefaultBrowser;
  } else if (notification_id == kNotificationIdSignIn) {
    return FeatureType::kSignIn;
  } else if (notification_id == kNotificationIdIncognitoTab) {
    return FeatureType::kIncognitoTab;
  } else if (notification_id == kNotificationIdNTPSuggestionCard) {
    return FeatureType::kNTPSuggestionCard;
  } else if (notification_id == kNotificationIdVoiceSearch) {
    return FeatureType::kVoiceSearch;
  }
  return FeatureType::kInvalid;
}

#if defined(OS_ANDROID)
base::Feature GetNotificationIphFeatureForFeature(FeatureType& feature) {
  switch (feature) {
    case FeatureType::kIncognitoTab:
      return feature_engagement::
          kIPHFeatureNotificationGuideIncognitoTabNotificationShownFeature;
    case FeatureType::kNTPSuggestionCard:
      return feature_engagement::
          kIPHFeatureNotificationGuideNTPSuggestionCardNotificationShownFeature;
    case FeatureType::kVoiceSearch:
      return feature_engagement::
          kIPHFeatureNotificationGuideVoiceSearchNotificationShownFeature;
    case FeatureType::kDefaultBrowser:
      return feature_engagement::
          kIPHFeatureNotificationGuideDefaultBrowserNotificationShownFeature;
    case FeatureType::kSignIn:
      return feature_engagement::
          kIPHFeatureNotificationGuideSignInNotificationShownFeature;
    default:
      NOTREACHED();
      return feature_engagement::
          kIPHFeatureNotificationGuideIncognitoTabNotificationShownFeature;
  }
}
#endif

}  // namespace feature_guide
