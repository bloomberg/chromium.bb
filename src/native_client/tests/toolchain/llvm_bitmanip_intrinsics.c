/*
 * Copyright (c) 2011 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * This test ensures that the backends can deal with bit manipulation
 * intrinsics.
 */

#include <stdio.h>
#include <stdint.h>
#include "native_client/tests/toolchain/utils.h"

/*  __builtin_popcount[ll](x) */
uint32_t llvm_intrinsic_ctpop(uint32_t) __asm__("llvm.ctpop.i32");
uint64_t llvm_intrinsic_ctpopll(uint64_t) __asm__("llvm.ctpop.i64");

/* __builtin_clz[ll](x) */
uint32_t llvm_intrinsic_ctlz(uint32_t, _Bool) __asm__("llvm.ctlz.i32");
uint64_t llvm_intrinsic_ctlzll(uint64_t, _Bool) __asm__("llvm.ctlz.i64");

/* __builtin_ctz[ll](x) */
uint32_t llvm_intrinsic_cttz(uint32_t, _Bool) __asm__("llvm.cttz.i32");
uint64_t llvm_intrinsic_cttzll(uint64_t, _Bool) __asm__("llvm.cttz.i64");

/* __builtin_bswap{16,32,64}(x) */
uint16_t llvm_intrinsic_bswap16(uint16_t) __asm__("llvm.bswap.i16");
uint32_t llvm_intrinsic_bswap32(uint32_t) __asm__("llvm.bswap.i32");
uint64_t llvm_intrinsic_bswap64(uint64_t) __asm__("llvm.bswap.i64");

/* volatile prevents partial evaluation by llvm */
volatile uint16_t i16[] = {0, 0xabcd, 0xdcba, 0xffff};
volatile uint32_t i32[] = {0, 0x01234567, 0x12345670, 0xffffffff};
volatile uint64_t i64[] = {0, 0x0123456789abcdefll,
                           0x0123456789abcdef0ll, 0xffffffffffffffffll};

/*
 * LLVM had a bug generating ARM code with large array offsets. The function
 * needs a constant offset > 255 to trigger the bug.
 */
volatile uint16_t i16l[257] = {[256] = 0xabcd};
volatile uint32_t i32l[257] = {[256] = 0x01234567};
volatile uint64_t i64l[257] = {[256] = 0x0123456789abcdefll};

#define print_op(op, x)                                                 \
  printf("%30s: %u\n", #op " (llvm)", (unsigned) llvm_intrinsic_ ## op(x))

#define print_op2(op, x, y)                                             \
  printf("%30s: %u\n", #op " (llvm)", (unsigned) llvm_intrinsic_ ## op(x, y))

#define print_op_builtin(op, x)                                     \
  printf("%30s: %u\n", #op " (builtin)", (unsigned) __builtin_ ## op(x))

int main(int argc, char* argv[]) {
  int i;

  for (i = 0; i < ARRAY_SIZE_UNSAFE(i16); ++i) {
    printf("\n%30s: 0x%04hx\n", "i16 value is", i16[i]);
    printf("%30s: 0x%04hx\n", "bswap (llvm)", llvm_intrinsic_bswap16(i16[i]));
    printf("%30s: 0x%04hx\n", "bswap (builtin)", __builtin_bswap16(i16[i]));
  }

  for (i = 0; i < ARRAY_SIZE_UNSAFE(i32); ++i) {
    printf("\n%30s: 0x%08x\n", "i32 value is", i32[i]);
    print_op(ctpop, i32[i]);
    print_op_builtin(popcount, i32[i]);

    print_op2(ctlz, i32[i], 0 /* Don't return undef for 0 input! */);
    /* 0 gives undefined results for the builtin on x86 */
    if (i32[i] == 0)
      printf("%30s: %u\n", "clz (builtin-manual-check)", 32);
    else
      print_op_builtin(clz, i32[i]);

    print_op2(cttz, i32[i], 0 /* Don't return undef for 0 input! */);
    /* 0 gives undefined results for the builtin on x86 */
    if (i32[i] == 0)
      printf("%30s: %u\n", "ctz (builtin-manual-check)", 32);
    else
      print_op_builtin(ctz, i32[i]);

    printf("%30s: 0x%08x\n", "bswap (llvm)", llvm_intrinsic_bswap32(i32[i]));
    printf("%30s: 0x%08x\n", "bswap (builtin)", __builtin_bswap32(i32[i]));
  }

  for (i = 0; i < ARRAY_SIZE_UNSAFE(i64); ++i) {
    printf("\n%30s: 0x%016llx\n", "i64 value is", i64[i]);
    print_op(ctpopll, i64[i]);
    print_op_builtin(popcountll, i64[i]);

    print_op2(ctlzll, i64[i], 0 /* Don't return undef for 0 input! */);
    /* 0 gives undefined results for the builtin on x86 */
    if (i64[i] == 0)
      printf("%30s: %u\n", "clzll (builtin-manual-check)", 64);
    else
      print_op_builtin(clzll, i64[i]);

    print_op2(cttzll, i64[i], 0 /* Don't return undef for 0 input! */);
    /* 0 gives undefined results for the builtin on x86 */
    if (i64[i] == 0)
      printf("%30s: %u\n", "ctzll (builtin-manual-check)", 64);
    else
      print_op_builtin(ctzll, i64[i]);

    printf("%30s: 0x%016llx\n", "bswapll (llvm)",
           llvm_intrinsic_bswap64(i64[i]));
    printf("%30s: 0x%016llx\n", "bswapll (builtin)",
           __builtin_bswap64(i64[i]));
  }

  printf("\nLarge offset tests:\n");
  printf("%30s: 0x%04hx\n", "bswap (llvm)", llvm_intrinsic_bswap16(i16l[256]));
  printf("%30s: 0x%04hx\n", "bswap (builtin)", __builtin_bswap16(i16l[256]));
  printf("%30s: 0x%08x\n", "bswap (llvm)", llvm_intrinsic_bswap32(i32l[256]));
  printf("%30s: 0x%08x\n", "bswap (builtin)", __builtin_bswap32(i32l[256]));
  printf("%30s: 0x%016llx\n", "bswapll (llvm)",
         llvm_intrinsic_bswap64(i64l[256]));
  printf("%30s: 0x%016llx\n", "bswapll (builtin)",
         __builtin_bswap64(i64l[256]));

  return 0;
}
