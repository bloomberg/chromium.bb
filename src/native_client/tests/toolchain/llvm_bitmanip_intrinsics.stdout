
                  i16 value is: 0x0000
                  bswap (llvm): 0x0000
               bswap (builtin): 0x0000

                  i16 value is: 0xabcd
                  bswap (llvm): 0xcdab
               bswap (builtin): 0xcdab

                  i16 value is: 0xdcba
                  bswap (llvm): 0xbadc
               bswap (builtin): 0xbadc

                  i16 value is: 0xffff
                  bswap (llvm): 0xffff
               bswap (builtin): 0xffff

                  i32 value is: 0x00000000
                  ctpop (llvm): 0
            popcount (builtin): 0
                   ctlz (llvm): 32
    clz (builtin-manual-check): 32
                   cttz (llvm): 32
    ctz (builtin-manual-check): 32
                  bswap (llvm): 0x00000000
               bswap (builtin): 0x00000000

                  i32 value is: 0x01234567
                  ctpop (llvm): 12
            popcount (builtin): 12
                   ctlz (llvm): 7
                 clz (builtin): 7
                   cttz (llvm): 0
                 ctz (builtin): 0
                  bswap (llvm): 0x67452301
               bswap (builtin): 0x67452301

                  i32 value is: 0x12345670
                  ctpop (llvm): 12
            popcount (builtin): 12
                   ctlz (llvm): 3
                 clz (builtin): 3
                   cttz (llvm): 4
                 ctz (builtin): 4
                  bswap (llvm): 0x70563412
               bswap (builtin): 0x70563412

                  i32 value is: 0xffffffff
                  ctpop (llvm): 32
            popcount (builtin): 32
                   ctlz (llvm): 0
                 clz (builtin): 0
                   cttz (llvm): 0
                 ctz (builtin): 0
                  bswap (llvm): 0xffffffff
               bswap (builtin): 0xffffffff

                  i64 value is: 0x0000000000000000
                ctpopll (llvm): 0
          popcountll (builtin): 0
                 ctlzll (llvm): 64
  clzll (builtin-manual-check): 64
                 cttzll (llvm): 64
  ctzll (builtin-manual-check): 64
                bswapll (llvm): 0x0000000000000000
             bswapll (builtin): 0x0000000000000000

                  i64 value is: 0x0123456789abcdef
                ctpopll (llvm): 32
          popcountll (builtin): 32
                 ctlzll (llvm): 7
               clzll (builtin): 7
                 cttzll (llvm): 0
               ctzll (builtin): 0
                bswapll (llvm): 0xefcdab8967452301
             bswapll (builtin): 0xefcdab8967452301

                  i64 value is: 0x123456789abcdef0
                ctpopll (llvm): 32
          popcountll (builtin): 32
                 ctlzll (llvm): 3
               clzll (builtin): 3
                 cttzll (llvm): 4
               ctzll (builtin): 4
                bswapll (llvm): 0xf0debc9a78563412
             bswapll (builtin): 0xf0debc9a78563412

                  i64 value is: 0xffffffffffffffff
                ctpopll (llvm): 64
          popcountll (builtin): 64
                 ctlzll (llvm): 0
               clzll (builtin): 0
                 cttzll (llvm): 0
               ctzll (builtin): 0
                bswapll (llvm): 0xffffffffffffffff
             bswapll (builtin): 0xffffffffffffffff

Large offset tests:
                  bswap (llvm): 0xcdab
               bswap (builtin): 0xcdab
                  bswap (llvm): 0x67452301
               bswap (builtin): 0x67452301
                bswapll (llvm): 0xefcdab8967452301
             bswapll (builtin): 0xefcdab8967452301
